### PR TITLE
ARROW-6983: [C++] Fix ThreadedTaskGroup lifetime issue

### PR DESCRIPTION
**Pull request by pitrou**
Together with a test.

### CODE REVIEW

File: cpp/src/arrow/util/task_group_test.cc [hunk 1]
```diff
@@ -212,6 +212,41 @@ void TestTasksSpawnTasks(std::shared_ptr<TaskGroup> task_group) {
   ASSERT_EQ(count.load(), (1 << (N + 1)) - 1);
 }
 
+// Try to replicate subtle lifetime issues when destroying a TaskGroup
+// where all tasks may not have finished running.
+void StressTaskGroupLifetime(std::function<std::shared_ptr<TaskGroup>()> factory) {
```
Comment (fsaintjacques) [inline cpp/src/arrow/util/task_group_test.cc]: I find that barrier-debugging is good to trigger such issues:

```
atomic<bool> barrier = false;
{
  TaskGroup group;

  group.Append([&barrier]) { while (!barrier) continue; /* or maybe do something with group? */ return Status::OK() });

  // Triggers group.Finish() to return early before all tasks are completed
  group.Append([]() {return Status::Invalid("oops")});

  group.Finish()
}

// Unlocks thread with group out of scope.
barrier = true;
```

Comment (pitrou) [reply cpp/src/arrow/util/task_group_test.cc]: I will try it, thank you.

Comment (pitrou) [reply cpp/src/arrow/util/task_group_test.cc]: Ok, it took some time, but I managed to write a failing test.

**Comment by pitrou** [comment]
@bkietz

**Comment by pitrou** [comment]
Travis-Ci: https://travis-ci.org/pitrou/arrow/builds/602350769
AppVeyor: https://ci.appveyor.com/project/pitrou/arrow/builds/28352718

**Review by bkietz — APPROVED** [review]

**Comment by wesm** [comment]
It's quite elegant solution here to creating a race condition on demand, may be useful in the future when such issues present themselves.

**Pull request closed without merging**